alternator/ttl.cc

```diff
@@ -261,6 +263,187 @@ static future<> expire_item(service::storage_proxy& proxy,
         qs.get_trace_state(), qs.get_permit());
 }
 
+// A class for iterating over all the token ranges *owned* by this shard.
+// We consider a token *owned* by this shard if:
+// 1. This node is a replica for this token.
+// 2. Moreover, this node is the *primary* replica of the token (i.e., the
+//    first replica in the ring).
+// 3. In this node, this shard is responsible for this token.
+// We will use this definition of which shard in the cluster owns which tokens
+// to split the expiration scanner's work between all the shards of the
+// system.
+// When some of the nodes in the system are down, nobody will be expiring
+// tokens owned by them, so we will need to separately detect when this
+// happens and let living nodes take over the work that a dead node cannot
+// do.
+// FIXME: need to decide how to choose primary ranges in multi-DC setup!
+// We could call get_primary_ranges_within_dc() below instead of get_primary_ranges().
+class token_ranges_owned_by_this_shard {
+    schema_ptr _s;
+    // _token_ranges will contain a list of token ranges owned by this node.
+    // We'll further need to split each such range to the pieces owned by
+    // the current shard, using _intersecter.
+    dht::token_range_vector _token_ranges;
+    dht::token_range_vector::const_iterator _range_it;
+    std::optional<dht::selective_token_range_sharder> _intersecter;
+public:
+    // FIXME: add a constructor to start iteration in the middle, to continue
+    // a scan after reboot.
```

Comment:
I don't think it's really important. Restarts are rare enough so that we can continue from the beginning. Perhaps we should start from a random location to avoid favoring one area.

Reply:
The random location idea is an interesting solution, indeed much better than starting at min_token.
Because the scan period can be long (e.g., 48 hours), it is not inconcievable that we can have a restart in the middle of such long period and interfere with the expiration delay guarantees the user asked for. I agree it's not critical in most cases.